### PR TITLE
Python: Fix schema building for complex types

### LINKED ISSUE #6388: Python: In v1.0.2 it's not possible to have lists or dictionaries as fields of a KernelBaseModel
**Describe the bug**
I am writing a native python plugin and I have this code 
```python
class Profile(KernelBaseModel):
    first_name: str
    last_name: str
    phone_numbers: list[str]

class ProfilePlugin:
    @kernel_function(name="GetPersonProfile", description="Gets a person's profile given their id.")
    def get_profile(self, person_id: str) -> Profile:
        return Profile(first_name="John", last_name="Doe")
```
My app fails to start up.  Stuff breaks because of the `phone_numbers`. If I comment out the `phone_numbers` then my app works fine.

Error:
```text
    if issubclass(parameter_type, KernelBaseModel):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen abc>", line 123, in __subclasscheck__
TypeError: issubclass() arg 1 must be a class
```

The problem is in the `KernelJsonSchemaBuilder`. The `issubclass` method is called with `list[str]`, which is not a class

This line:
`if issubclass(parameter_type, KernelBaseModel):`

should be replaced with:
`if isinstance(parameter_type, KernelBaseModel):`

**To Reproduce**
Described above.

**Expected behavior**
Described above.

**Screenshots**
NA

**Platform**
 - OS: [Mac]
 - IDE: [VS Code]
 - Language: [Python]
 - Source: [v1.0.2]

**Additional context**
NA

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

The schema building for complex types worked in some regard, but it didn't work for all cases. There needs to be better checks on complex types, if they have __args__ and if so, recurse to the base class type, and also grab the type that is specified in the list like list[CustomClass], list[str] or even Union[int, bool]. 

### Description

This PR improves the parsing/building of complex types. Added integration tests to cover more types in the hope that this doesn't lead to any more issues.
- Closes #6388 

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/d31be32b75dc56e00f474db90d5fbb98c8ebbb50/README.md">Py3.10 Test Coverage</a><details><summary>Python 3.10 Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td><b>TOTAL</b></td><td><b>6096</b></td><td><b>879</b></td><td><b>86%</b></td><td>&nbsp;</td></tr></tbody></table><i>report-only-changed-files is enabled. No files were changed during this commit :)</i></details>

## Python 3.10 Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 1374 | 1 :zzz: | 0 :x: | 0 :fire: | 13.340s :stopwatch: |

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
Mostly testing, but forward refs vs regular definition can be a pain to deal with!

**Review by eavanvalkenburg — APPROVED** [review]
LGTM

File: python/semantic_kernel/schema/kernel_json_schema_builder.py [hunk 1]
```diff
@@ -69,7 +101,54 @@ def build_from_type_name(cls, parameter_type: str, description: str | None = Non
...
+            _, value_type = args
+            additional_properties = cls.build(value_type)
+            if additional_properties == {"type": "object"}:
+                additional_properties["properties"] = {}  # Account for differences in Python 3.10 dict
+            return {"type": "object", "additionalProperties": additional_properties, "description": description}
+        if origin is tuple:
+            items = [cls.build(arg) for arg in args]
+            return {"type": "array", "items": items, "description": description}
+        if origin is Union:
+            if len(args) == 2 and type(None) in args:
```
Comment (alliscode) [inline python/semantic_kernel/schema/kernel_json_schema_builder.py:143]: More comments could be helpful in some of this code. For example, future you and current me would be curious why we special case the len(args) == 2 case.

Comment (moonbox3) [reply python/semantic_kernel/schema/kernel_json_schema_builder.py:143]: Great point, let me add.

**Pull request merged**